compiler/fir/checkers/checkers.jvm/src/org/jetbrains/kotlin/fir/analysis/jvm/checkers/expression/FirJavaGenericVarianceViolationTypeChecker.kt

```diff
@@ -0,0 +1,187 @@
+            // same.
+            if (lowerConstructor == upperConstructor) continue
+
+            // If the base class of the argument type is not equal or a sub class of the lower bound, then we simply allow it so that
+            // Kotlin immutable collections can be used in place of Java collection types.
+            if (!typeCtx.isTypeConstructorEqualOrSubClassOf(argType, lowerBound)) continue
+
+            // In general, out type projection makes a mutable collection "readonly". A priori such a projected type is not a subtype of a
+            // non-projected type because projection has "removed" the ability to write to this collection. But for the purpose of this
+            // checker, we don't care about the readability/writability of collections.
+            //
+            // More importantly, removing such out projection is a
+            // shortcut to make this type comparable with the expected type by the function (after removing type capturing below). Consider
+            // the code below (here we use kotlin code in the example, but the same idea applies to calls to Java methods as well)
+            //
+            // ```
+            // fun get(): MutableList<out String> = ...
+            // fun <T> take(l: MutableList<T>) = ...
+            // fun test() {
+            //   take(get())
+            // }
+            // ```
+            //
+            // The type of `get()` is `MutableList<out String>`. After type parameter instantiation, `take` takes a
+            // `MutableList<Captured<out String>>`. Obviously, a `MutableList<out String>` is not a subtype of `MutableList<Captured<out
+            // String>>` because we have lost the identity of the captured type at this point: we no longer know that the captured type is
+            // actually created because of type projection from `get`. Hence, to workaround this problem, we simply remove all the out
+            // projection and type capturing and compare the types after such erasure. This way, we won't incorrectly reject any valid code
+            // though we may accept some invalid code. But in presence of the unsound flexible types, we are allowing invalid code already.
+            val argTypeWithoutOutProjection = argType.removeOutProjection(true)
+            val lowerBoundWithoutCapturing = context.session.inferenceComponents.approximator.approximateToSuperType(
+                lowerBound,
+                TypeApproximatorConfiguration.FinalApproximationAfterResolutionAndInference
+            ) ?: lowerBound
+
+            if (!AbstractTypeChecker.isSubtypeOf(
+                    typeCtx,
+                    argTypeWithoutOutProjection,
+                    lowerBoundWithoutCapturing.withNullability(ConeNullability.NULLABLE, typeCtx)
+                )
+            ) {
+                reporter.reportOn(arg.source, FirErrors.JAVA_TYPE_MISMATCH, argType, expectedType, context)
```

Comment:
Looks like `argType` and `expectedType` should be swapped

Reply:
There's still an error:
`e: /testProject/idea/src/org/jetbrains/kotlin/idea/testIntegration/KotlinTestFinder.kt: (86, 59): Java type mismatch expected java/util/ArrayList<ft<com/intellij/openapi/util/Pair<out com/intellij/psi/PsiNamedElement, kotlin/Int>, com/intellij/openapi/util/Pair<out com/intellij/psi/PsiNamedElement, kotlin/Int>?>> but found ft<@EnhancedNullability kotlin/collections/MutableCollection<ft<com/intellij/openapi/util/Pair<out com/intellij/psi/PsiNamedElement, kotlin/Int>, com/intellij/openapi/util/Pair<out com/intellij/psi/PsiNamedElement, kotlin/Int>?>>, @EnhancedNullability kotlin/collections/Collection<ft<com/intellij/openapi/util/Pair<out com/intellij/psi/PsiNamedElement, kotlin/Int>, com/intellij/openapi/util/Pair<out com/intellij/psi/PsiNamedElement, kotlin/Int>?>>>. Use explicit cast`

It happens when compiling old version of our project: https://github.com/JetBrains/kotlin/blob/1.5.30/idea/src/org/jetbrains/kotlin/idea/testIntegration/KotlinTestFinder.kt#L86
While the code should be green

Reply:
Unfortunately, these tests are failing in your branch:
DiagnosticTestGenerated$Tests$Generics$TpAsReified.testCapturedAsReified 
FirOldFrontendDiagnosticsTestGenerated$Tests$Generics$TpAsReified.testCapturedAsReified
FirBlackBoxInlineCodegenTestGenerated$Signature.testInProjectionSubstitution

Reply:
Oops, my bad! I missed changing https://github.com/JetBrains/kotlin/blob/c2cf2f36cd6ce9236d8a30f164bca229847f2a5c/compiler/resolution.common/src/org/jetbrains/kotlin/types/TypeApproximatorConfiguration.kt#L89-L88